web/packages/extension/src/player.ts

```diff
@@ -236,22 +239,120 @@ window.addEventListener("load", () => {
+            return swfUrl;
+        }
+    };
 
     const options = await utils.getExplicitOptions();
     localFileName.textContent = document.title;
     localFileInput.value = "";
-    load({
-        ...options,
-        url: swfUrl,
-        base: swfUrl.substring(0, swfUrl.lastIndexOf("/") + 1),
-        ...baseExtensionConfig,
-    });
+    load(
+        {
+            ...options,
+            url: swfUrl,
+            base: swfUrl.substring(0, swfUrl.lastIndexOf("/") + 1),
+            ...baseExtensionConfig,
+        },
+        resolveUrlHook,
+    );
+}
+
+/**
+ * Resolves a given string to a URL if possible.
+ * If the protocol is missing and the string is otherwise a valid web URL, https:// is inserted if the
+ * server supports https, otherwise http.
+ * If the protocol is missing and the string is otherwise a valid file URL, file:/// is inserted.
+ * If the string can't be resolved to a URL, null is returned.
+ * @param enteredUrl The string that should be resolved to a URL.
+ * @return The resolved URL object.
+ */
+async function resolveSwfUrl(enteredUrl: string): Promise<URL | null> {
+    // TODO: Use canParse in the future when it doesn't break browser compatibility
+    // If the URL is (very likely) a file URL with missing file protocol, we return it as file URL
+    // Must be the first test as the URL constructor accepts C:\… as URL with protocol C
+    if (enteredUrl.match(/^[A-Za-z]:\\|^[/~\\]/)) {
+        try {
+            return new URL("file:///" + enteredUrl);
+        } catch (error) {
+            return null;
+        }
+    }
+
+    try {
+        return new URL(enteredUrl);
+    } catch (error) {
+        // The protocol is missing
+
+        // If the URL doesn't contain a dot, it can't be a valid web URL
+        // The URL constructor doesn't check this if a protocol exists
+        if (!enteredUrl.includes(".")) {
+            return null;
+        }
+
+        try {
+            // Only use http if it works but https doesn't
+            if (
+                (await serverAvailable("https://" + enteredUrl, 600)) ||
+                !(await serverAvailable("http://" + enteredUrl, 300))
+            ) {
+                return new URL("https://" + enteredUrl);
+            } else {
+                return new URL("http://" + enteredUrl);
+            }
+        } catch (error) {
+            return null;
+        }
+    }
+}
+
+/**
+ * Tests and returns whether a server exists under a given URL.
+ * @param url The URL that should be tested.
+ * @param timeout The maximum number of milliseconds that should be waited for a response.
+ * @return Whether a server exists under the given URL.
+ */
+async function serverAvailable(
+    url: string | URL,
+    timeout: number,
+): Promise<boolean> {
+    function timeoutFetch(
+        url: string | URL,
+        options: RequestInit,
+        timeout: number,
+    ) {
+        // TODO: Use AbortSignal in the future when it doesn't break browser compatibility
```

Comment:
Is it really too new?

Reply:
Yes, it breaks our current [browser compatibility](https://github.com/ruffle-rs/ruffle/wiki)

Reply:
I think we could do `await Promise.race([fetch(), new Promise((_, reject) => setTimeout(reject, timeout))])` maybe?

Reply:
Actually, shouldn't we support it?
https://developer.mozilla.org/en-US/docs/Web/API/Request/signal says
Chrome 66
Firefox 57
Safari 12.1

Or are you hoping to use `AbortSignal.timeout` directly?

Reply:
I meant something like `fetch(url, { signal: AbortSignal.timeout(…) })`. But its [minimum supported version](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/timeout_static) is over ours.
But you are right, [AbortController()](https://developer.mozilla.org/en-US/docs/Web/API/AbortController/AbortController), [AbortController#abort()](https://developer.mozilla.org/en-US/docs/Web/API/AbortController/abort) and [Request#signal](https://developer.mozilla.org/en-US/docs/Web/API/Request/signal) are all available for our minimum supported versions, so we could use a polyfill like
```
AbortSignal.timeout ??= function timeout(ms) {
  const ctrl = new AbortController()
  setTimeout(() => ctrl.abort(), ms)
  return ctrl.signal
}
```
([Reference](https://stackoverflow.com/questions/46946380/fetch-api-request-timeout/57888548))
I'll change that :)